clients/drcachesim/scheduler/flexible_queue.h

```diff
@@ -0,0 +1,219 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* A priority queue with constant-time search and removal from the middle. */
+
+#ifndef _DRMEMTRACE_FLEXIBLE_QUEUE_H_
+#define _DRMEMTRACE_FLEXIBLE_QUEUE_H_ 1
+
+/**
+ * @file drmemtrace/flexible_queue.h
+ * @brief DrMemtrace flexible priority queue.
+ */
+
+#define NOMINMAX // Avoid windows.h messing up std::max.
+#include <iostream>
+#include <limits>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "utils.h"
+
+namespace dynamorio {  /**< General DynamoRIO namespace. */
+namespace drmemtrace { /**< DrMemtrace tracing + simulation infrastructure namespace. */
+
+/**
+ * A priority queue with constant-time search and removal from the middle.
+ */
+template <typename T, class comparator_t = std::less<T>> class flexible_queue_t {
+public:
+    typedef typename std::vector<T>::size_type index_t;
+    // Wrap max in parens to work around Visual Studio compiler issues with the
+    // max macro (even despite NOMINMAX defined above).
+    static constexpr index_t INVALID_INDEX = (std::numeric_limits<index_t>::max)();
+
+    flexible_queue_t() = default;
+    explicit flexible_queue_t(int verbose)
+        : verbose_(verbose)
+    {
+    }
+    bool
+    push(T entry)
+    {
+        if (entry2index_.find(entry) != entry2index_.end())
+            return false; // Duplicates not allowed.
+        entries_.push_back(entry);
+        index_t node = entries_.size() - 1;
+        entry2index_[entry] = node;
+        percolate_up(node);
+        print("after push");
```

Comment:
That was my original plan as it is a nop by default but it does have a function call cost.  Maybe it should be disabled more.  Still, it may be useful in the future.  I renamed it to `vprint` with a verbosity threshold so it's clearer what it is.